src/state/api/__tests__/apiSlice.test.ts

```diff
@@ -0,0 +1,95 @@
+import { rest } from 'msw'
+import { vi } from 'vitest'
+import apiSlice from '../apiSlice'
+
+import { server } from '../../../api/server'
+const pensjonsberegningData = require('../../../api/__mocks__/pensjonsberegning.json')
+
+import { setupApiStore } from '../../../test-utils'
+
+vi.mock('@reduxjs/toolkit/query/react')
+
+describe('apiSlice', async () => {
```

Comment:
Trenger vi egentlig disse testene?

De er veldig tidkrevende å sette opp, bruker veldig mye mocking og tester ikke koden slik den faktisk brukes, ref https://redux.js.org/usage/writing-tests:

> Prefer writing integration tests with everything working together. For a React app using Redux, render a <Provider> with a real store instance wrapping the components being tested. Interactions with the page being tested should use real Redux logic, with API calls mocked out so app code doesn't have to change, and assert that the UI is updated appropriately.

En annen relevant quote fra de offisielle docsa:

> Do not try to mock selector functions or the React-Redux hooks! Mocking imports from libraries is fragile, and doesn't give you confidence that your actual app code is working.

Her mener jeg det kan være verdt å heller teste endepunktene ved å mounte komponentene som bruker de og så teste at vi rendrer ønsket markup.

Reply:
Om vi har veldig kompliserte selectorer eller reducers tenker jeg at de godt kan testes i isolasjon

Reply:
Sorry for en vegg av tekst 😅 Jeg syns det er kjempebra at vi utfordrer hverandre på disse spørsmålene, og jeg håper ikke svaret mitt her blir tatt på noen negativ måte.

> Som sagt, den ene utelukker ikke den andre. Disse er testene som skrives en gang (og det er gjort nå, så du får det gratis :) ). De er enhetstester og låser koden på et annet nivå enn integrasjonstestene ville gjøre.

Ingen kode er gratis, og det gjelder også tester. Tester må vedlikeholdes på samme måte som annen kode og har en kostnad i utviklertid. Å ikke teste unødvendig er et mål i seg selv, så får man heller diskutere hva som er unødvendige tester.

> de er eneste måte å sjekke på at selve setup av slicen er riktig. At path, env variabler, og oppsett mot api'et i prod er riktig. Det er som sagt en one-shot'er som "låser" det at vårt oppsett virker slik den skal.

Det jeg prøver å utfordre er ideen om at vi trenger å eksplisitt teste at slicen i seg selv er satt opp riktig. Det blir jo uansett testet implisitt ved at i dette tilfellet `Pensjonsberegning`-testen tar for seg de forskjellige responsene vi kan motta fra backend, så det er da mao. ikke den eneste måten å teste at slicen er satt opp riktig. Cypress har også en smoketest hvis hensikt er å sjekke at alt av config er satt opp riktig og at appen kjører. Hvis noen endrer på oppsettet i slicen, f.eks. ved å endre pathen, vil disse testene feile. Det jeg syns er interessant å teste er forskjellige situasjoner som kan oppstå mens en bruker interagerer med appen og hvordan vi velger å vise dette i UIet. Alt annet er egentlig implementasjonsdetaljer siden brukeren ikke har noe forhold til det. Dersom valideringslogikken i slicen er kjempekomplisert kan man kanskje teste valideringsfunksjonen i isolasjon, men når man trenger å mocke halve verden for å få testene til å kjøre syns jeg det gir oss mer smerte enn verdi å teste på dette nivået, og det gir iaf. ikke meg noen større trygghet på at appen vår funker som den skal.

> de er også eneste måte til å sjekke hvordan responsen bearbeides både ved success og feil. Denne logikken er uavhengig av hva komponenten gjør med den. Det ene er det som mottas fra backend, det andre er hva vi gjør med det og lagrer i vår store. Det tredje er hva komponenten velger å vise. Uten enhetstester kan vi ikke ha kontroll på biten i midten og bla logikken rundt vår typeguard.

Hvordan responsen bearbeides i selve oppsettet av slicen er en implementasjonsdetalj på samme måte som forklart over. Jeg skjønner at logikken i slicen er uavhengig fra komponenten som bruker den, men jeg mener at det er hvordan komponenten bruker den som faktisk er interessant siden det er dette som er merkbart for bruker. Hvis logikken er kjempekomplisert kan jeg se verdien av å teste det i isolasjon gitt at man kan gjøre det på en måte som bruker mocking i minst mulig grad. Typeguards er veldig lette å teste i isolasjon og det kan man godt gjøre.

> for at coverage skal ha noe verdi, må vi faktisk prøve å oppnå den. Testing av hver branch og functions er en forutsetning for å kunne trace og enklere debugge når kompleksiteten vil øke

Coverage i seg selv har liten verdi. Det sier oss ingenting om hva som blir testet eller hvor god kvalitet testene har. Det eneste det forteller oss er hvilke kodelinjer som har kjørt under kjøring av tester. Det kan være deler av applikasjonen som _ikke_ er testet med fornuftige assertions men som ser ut til å være testet siden linjene har blitt kjørt under en test. Det kan være en interessant metrikk om man bruker det for å finne ut av hvor man definitivt kan skrive flere tester, men jeg mener at en høy coverage ikke bør være et mål i seg selv.